include/shark/Algorithms/Trainers/CARTcommon.h

```diff
@@ -0,0 +1,156 @@
+#define SHARK_ALGORITHMS_TRAINERS_CARTCOMMON_H
+
+#include <shark/Models/Trees/General.h>
+#include <shark/Data/DataView.h>
+#include <vector>
+#include <unordered_map>
+#include <utility>
+namespace shark {
+namespace detail {
+namespace cart {
+/// ClassVector
+using ClassVector = UIntVector;
+
+
+struct Attribute {
+	double value;
+	std::size_t id;
+
+	inline bool operator<(Attribute const &v) const {
+		return value < v.value;
+	}
+};
+
+
+class Index {
+/// attribute table
+	using AttributeTable = std::vector<Attribute>;
+/// collecting of attribute tables
+	using AttributeTables = std::vector<AttributeTable>;
+
+	std::size_t m_noInputDimensions, m_noElements;
+	AttributeTables m_tables;
+
+	explicit Index(AttributeTables &&tables)
+			: m_noElements(tables[0].size()),
+			  m_noInputDimensions(tables.size()),
+			  m_tables(std::move(tables))
+	{}
+public:
+	struct size_pair{
+		std::size_t n_inputDimensions, n_elements;
+		size_pair(std::size_t n_inputDimensions, std::size_t n_elements)
+				: n_inputDimensions(n_inputDimensions),
+				  n_elements(n_elements)
+		{}
+	};
+
+/** Creates an index of the dataset
+ *  A dataset consisting of m input variables has m sorted attribute tables.
+ *  [attribute | rid ]
+ */
+	template<class Dataset>
+	explicit Index(DataView<Dataset const> const &elements)
+			: m_noElements{elements.size()},
+			  m_noInputDimensions{elements[0].input.size()},
+			  m_tables(m_noInputDimensions)
+	{
+		std::size_t n_elements = m_noElements;
+		//Each entry in the outer vector is an attribute table
+		//For each column
+		for (std::size_t j = 0; j < m_noInputDimensions; j++) {
+			//m_tables.push_back({});
+			auto &table = m_tables[j];
+			table.reserve(n_elements);
+
+			//For each row, store Attribute value, class and rowId
+			for (std::size_t i = 0; i < n_elements; i++) {
+				table.push_back(Attribute{elements[i].input[j], i});
+			}
+			std::sort(table.begin(), table.end());
+		}
+	}
+
+/**
+ * Returns two Indices: left and right
+ * Calculated from splitting tables at (index, valIndex)
+ */
+	std::pair<Index, Index> split(std::size_t index, std::size_t valIndex) {
+		//Build a hash table for fast lookup
+		std::unordered_map<std::size_t, bool> hash;
+		for(std::size_t i = 0, s = m_tables[index].size(); i<s; ++i) {
+			hash[m_tables[index][i].id] = (i<=valIndex);
+		}
+
+		AttributeTables RAttributeTables, LAttributeTables;
+		for(auto && table : m_tables) {
+			auto begin = table.begin(), end = table.end();
+			auto middle = std::stable_partition(begin,end,[&hash](Attribute const& entry){
+				return hash[entry.id];
+			});
+			RAttributeTables.emplace_back(AttributeTable{middle,end});
+			table.resize(std::distance(begin,middle));
+			LAttributeTables.emplace_back(std::move(table));
+		}
+		m_tables.clear(); m_tables.shrink_to_fit();
+		return std::make_pair(Index{std::move(LAttributeTables)},
+							  Index{std::move(RAttributeTables)});
+	}
+
+	inline size_pair size() const {
```

Comment:
i would use two different functions here. size() has a very special name in c++, not to tinker with :)